rebasehelper/tags.py

```diff
@@ -84,26 +116,38 @@ def sanitize(name):
                     return '{0}0'.format(tokens[0])
                 return '{0}{1}'.format(tokens[0], int(tokens[1]))
             return name.capitalize()
-        result = []
+        result: List[Tag] = []
         tag_re = re.compile(r'^(?P<prefix>(?P<name>\w+)\s*:\s*)(?P<value>.+)$')
         parsed_idx = 0
-        for section, section_content in raw_content.sections:
+        for section_index, (section, section_content) in enumerate(raw_content.sections):
             section = section.lower()
-            if not section.startswith('%package'):
+            is_list = section.startswith('%sourcelist') or section.startswith('%patchlist')
+            if not section.startswith('%package') and not is_list:
                 continue
             if parsed_idx >= len(parsed_content.sections) or section != parsed_content.sections[parsed_idx][0].lower():
                 parsed: List[str] = []
             else:
                 parsed = parsed_content.sections[parsed_idx][1]
                 parsed_idx += 1
+            if section.startswith('%sourcelist'):
+                # copy the list, we do not want to permanently modify the content
+                section_content = cls._prepend_list_with_tags(section, section_content[:], parsed, 'Source',
+                                                              cast(Tuple[Tag], tuple(result)))
+            elif section.startswith('%patchlist'):
+                section_content = cls._prepend_list_with_tags(section, section_content[:], parsed, 'Patch',
+                                                              cast(Tuple[Tag], tuple(result)))
             for index, line in enumerate(section_content):
                 expanded = MacroHelper.expand(line)
                 if not line or not expanded:
                     continue
-                valid = bool(parsed and [p for p in parsed if p == expanded.rstrip()])
+                # There doesn't seem to be a way to parse conditionalized tags from sourcelist
+                # and patchlist. Just ignore such tags and consider only valid ones (this gets
+                # taken care of in _prepend_list_with_tags).
```

Comment:
I don't understand this justification at all. Can you elaborate?

Reply:
There is an example in the text of this PR. Lets say that the section looks like this:

```
%sourcelist
source2.tar.gz
%if 0%{?source:1}
source3.tar
%endif
```

Obviously if the macro `source` is not defined, then the condition won't be satisfied and the source won't be in parsed spec content. Currently Tags also parse conditionalized macros in preamble but I don't think that the same can be done for %sourcelist/%patchlist. We can't automatically prepend all lines because then the `%if` statement would be counted as Source. But then there is no way to parse `source3.tar` as a `Source` tag

Reply:
> Obviously if the macro `source` is not defined, then the condition won't be satisfied and the source won't be in parsed spec content.

Exactly, that's the whole point.

> We can't automatically prepend all lines because then the `%if` statement would be counted as Source.

Then just preprend only the lines that are in `parsed`.

Reply:
But I can see this prepending thing is getting more and more complicated, maybe it would be better to forget about it and just special case the list sections and implement their own parsing.

Reply:
> Then just preprend only the lines that are in `parsed`. Or, which would be probably better, only prepend the lines that don't expand to nothing.

That's what I am doing. I was just trying to point out that unlike with tags in `%package` sections (where even conditionalized tags are parsed but with `valid=False`), the tags that aren't in `parsed` are ignored. I think that both should be use (check whether it actually expands to something and whether it is in `parsed`).

As for your second point - I thought about that and even started working on a solution before starting the "prepend" approach. I don't think that it would help. If we wanted to have a robust parsing, it would result in a lot of code duplication which would be quite hard to convert into common logic (at least from what I've tried)

Reply:
> where even conditionalized tags are parsed but with `valid=False`

That's not the case for tags defined in conditional macros though. Those are ignored as well.

> If we wanted to have a robust parsing, it would result in a lot of code duplication which would be quite hard to convert into common logic (at least from what I've tried)

Yes, there would be a lot of duplicated code, but isn't that better than having a single codebase but with special-casing almost everywhere?

Reply:
It probably is. Also it won't be that bad considering all of the above. We can keep the code for `%package` sections as it is and for `%sourcelist`/`%patchlist` it should just be a matter of going through all the lines and creating tags from the valid lines.